### ISSUE TITLE
Error wrappiing GYM environment into SubprocVecEnv object

### ISSUE THREAD

**Issue by EloyAnguiano** [rte-france/Grid2Op#431]
## Environment
 - Grid2op version: `1.7.1`
 - System: `ubuntu20.04`
 - Additional system information
 - Stable baselines 3 version: 1.7.0
 - Python version: 3.10.4

## Bug description
When I try to wrap the converted gym envioronment into a SubprocVecEnv object from SB3 (for efficiency), It returns what it seems to be a pickling error.

### Code snippet

```python
from copy import deepcopy

from grid2op import make
from grid2op.gym_compat import DiscreteActSpace, GymEnv
from stable_baselines3.common.vec_env import SubprocVecEnv

if __name__ == "__main__":

    grid2op_env = make("rte_case14_redisp")
    gym_env = GymEnv(grid2op_env)  # create the gym environment
    gym_env.action_space = DiscreteActSpace(grid2op_env.action_space)

    gym_env = SubprocVecEnv([deepcopy(gym_env) for _ in range(5)])
    print("DONE")
```

## Current output
```
AttributeError: type object 'Environment_rte_case14_redisp' has no attribute 'process_grid2op_shunt_data'. Did you mean: 'process_grid2op_compat'?
Traceback (most recent call last):
  File "/home/eloy.anguiano/miniconda3/envs/l2rpn_new/lib/python3.10/multiprocessing/forkserver.py", line 274, in main
    code = _serve_one(child_r, fds,
  File "/home/eloy.anguiano/miniconda3/envs/l2rpn_new/lib/python3.10/multiprocessing/forkserver.py", line 313, in _serve_one
    code = spawn._main(child_r, parent_sentinel)
  File "/home/eloy.anguiano/miniconda3/envs/l2rpn_new/lib/python3.10/multiprocessing/spawn.py", line 126, in _main
    self = reduction.pickle.load(from_parent)
  File "/home/eloy.anguiano/miniconda3/envs/l2rpn_new/lib/python3.10/site-packages/stable_baselines3/common/vec_env/base_vec_env.py", line 375, in __setstate__
    self.var = cloudpickle.loads(var)
  File "/home/eloy.anguiano/miniconda3/envs/l2rpn_new/lib/python3.10/site-packages/grid2op/Space/GridObjects.py", line 3742, in init_grid_from_dict_for_pickle
    res_cls.process_grid2op_shunt_data()
```

## Expected output
```
DONE
```

**Comment by BDonnot** [comment]
Hello,

This specific error is related to an old grid2op version. It has been fixed in grid2op 1.8.0 (see changelog for more information). So I suggest you upgrade to this version. Latest version is probably a good idea :-)

But, once you do that, you will most likely run into another trouble with pickle... For that you have 2 solutions:

# Solution 1: use the 'experimental_read_from_local_dir' feature

This feature is experimental, but each time I tested it it works.

It works in two steps.

FIrst step you generate the classes and save them to the disk as regular python file. You can do so with `env.generate_classes()`: https://grid2op.readthedocs.io/en/latest/environment.html#grid2op.Environment.BaseEnv.generate_classes

This first step is to do only once.

And then, you add the key word `experimental_read_from_local_dir=True` to the call to `grid2op.make(...)` and then you won't have any trouble with pickle (probided that you don't change the backend for example, if you do you'll have to redo step 1 above)

# Solution 2: small hack

You can get inspired from this code, which worked on my machine:

```python
from stable_baselines3.common.env_util import make_vec_env
import gym
import grid2op
from grid2op.gym_compat import GymEnv, BoxGymObsSpace, DiscreteActSpace, BoxGymActSpace
from gym.spaces import Discrete, Box
from stable_baselines3 import PPO
from stable_baselines3.common.vec_env import DummyVecEnv, SubprocVecEnv


class GymEnvMultiProc(gym.Env):
    def __init__(self, gym_env_g2op):
        self.gym_env_g2op = gym_env_g2op
        self.action_space = Discrete(n=self.gym_env_g2op.action_space.n)
        self.observation_space = Box(low=self.gym_env_g2op.observation_space.low, 
                                     high=self.gym_env_g2op.observation_space.high)
    def step(self, *args, **kwargs):
        return self.gym_env_g2op.step(*args, **kwargs)
    
    def reset(self, *args, **kwargs):
        return self.gym_env_g2op.reset(*args, **kwargs)
    
    def seed(self, *args, **kwargs):
        return self.gym_env_g2op.seed(*args, **kwargs)

    
def make_env():
    env = grid2op.make("l2rpn_case14_sandbox")
    gym_env = GymEnv(env)
    gym_env.observation_space = BoxGymObsSpace(env.observation_space, attr_to_keep=["rho"])
    gym_env.action_space = DiscreteActSpace(env.action_space, attr_to_keep=["set_bus"])
    env_ok = GymEnvMultiProc(gym_env)
    return env_ok


if __name__ == '__main__':
    env = make_vec_env(make_env, n_envs=16, vec_env_cls=SubprocVecEnv)
    model = PPO("MlpPolicy", env, verbose=1)
    model.learn(total_timesteps=250)
```

**Comment by BDonnot** [comment]
Please consider using one of two "solutions" I gave in my answer. 

It appears you used neither

**Comment by EloyAnguiano** [comment]
I did what you suggested at the first solution you provided:
```
    grid2op_env = make("rte_case14_redisp")
    grid2op_env.generate_classes()
    grid2op_env = make("rte_case14_redisp", experimental_read_from_local_dir=True)
```
Here I `make` the environment two times to check later if any change to the configuration of both `makes` could be made.

**Comment by BDonnot** [comment]
Maybe I did not explain the first solution correctly then. Sorry about that. 

It works in two steps.

# Step 1 (do it only once)

First step you generate the classes and save them to the disk as regular python file. You can do so with env.generate_classes(): https://grid2op.readthedocs.io/en/latest/environment.html#grid2op.Environment.BaseEnv.generate_classes

This first step is to do only once, in a separate script then.

# Step 2 (as many times as you want)

Load your env with the key-word argument `experimental_read_from_local_dir=True`

**NB** if you do both steps in the same script it will NOT work.

Also, you might consider using:
- MultiFolderWithCache (see the doc https://grid2op.readthedocs.io/en/latest/environment.html#optimize-the-data-pipeline) to not spend the majority of your time reading data from the hard drive
- lightsim2grid as `backend` to be able to perform steps up to 15x faster for the same exact output (up to numerical precision) [cf. https://github.com/BDonnot/lightsim2grid/#1-as-a-grid2op-backend-preferred-method]

**Comment by BDonnot** [comment]
> PD1: Using lightsim2grid as backend returned an error sometimes when I tried to rerun a script.

Let me know it's important I can find the bugs and fix them. You are probably not alone to encounter them (also these bugs might appear with lightsim because it's faster so the "time to get the bug" is 10x faster with lightsim this is why you don't notice it with pandapower: you didn't "wait" enough)

> Have you located the part that is giving problems? 

I know exactly where the "pickle" problem is and I'm not sure anyone can do anything about it. But maybe i'm wrong and I would appreciate any help regarding the matter.

One of the reason grid2op is fast is that I modify the class (not the instance of the class, but the class) depending on the environment. This is why if you simply print the type of your env, you will see something that ends with the environment name. 

This allows to not "copy" the grid at each step and to access things like `obs.n_line` or `env.n_line` or `env.action_space.n_line` and whole lot of functions and attributes.

The fact you need to run the "generate classes" in a different script is because python will define the classes only once (the first call to make) and then re use them if you load the same environment. So basically in your first script the "experimental_read_from_local_dir" has absolutely no impact as the class are already generated.

And this causes trouble with pickle because the way pickle works. It requires to find the definition of the class before unpickling some object. Here the class are precisely generated when the first object is created (and not after).

What the "generate_classes()" does is simply write the definition of the class into a folder. And then the "read_from_local_dir" load the class from there instead of generating the class. This way pickle is "happy" because it knows where to find the definition (but again if you know a way to force pickle to execute a code when it sees a type it does not know then let me know)

> Maybe I can help you with that so you don't have to relegate in disk steps and forced script changes.

For that I would need to "send classes" to pickle without relying on disk. I looked for ways to do that but did not find anything suitable. I'd be happy to have a better solution indeed.

**Comment by BDonnot** [comment]
For your usecase, you might prefer to consider the other option I mentionned at the beginning: 

# Second solution
You can get inspired from this code, which worked on my machine:

```python
from stable_baselines3.common.env_util import make_vec_env
import gym
import grid2op
from grid2op.gym_compat import GymEnv, BoxGymObsSpace, DiscreteActSpace, BoxGymActSpace
from gym.spaces import Discrete, Box
from stable_baselines3 import PPO
from stable_baselines3.common.vec_env import DummyVecEnv, SubprocVecEnv


class GymEnvMultiProc(gym.Env):
    def __init__(self, gym_env_g2op):
        self.gym_env_g2op = gym_env_g2op
        self.action_space = Discrete(n=self.gym_env_g2op.action_space.n)
        self.observation_space = Box(low=self.gym_env_g2op.observation_space.low, 
                                     high=self.gym_env_g2op.observation_space.high)
    def step(self, *args, **kwargs):
        return self.gym_env_g2op.step(*args, **kwargs)
    
    def reset(self, *args, **kwargs):
        return self.gym_env_g2op.reset(*args, **kwargs)
    
    def seed(self, *args, **kwargs):
        return self.gym_env_g2op.seed(*args, **kwargs)

    
def make_env():
    env = grid2op.make("l2rpn_case14_sandbox")
    gym_env = GymEnv(env)
    gym_env.observation_space = BoxGymObsSpace(env.observation_space, attr_to_keep=["rho"])
    gym_env.action_space = DiscreteActSpace(env.action_space, attr_to_keep=["set_bus"])
    env_ok = GymEnvMultiProc(gym_env)
    return env_ok


if __name__ == '__main__':
    env = make_vec_env(make_env, n_envs=16, vec_env_cls=SubprocVecEnv)
    model = PPO("MlpPolicy", env, verbose=1)
    model.learn(total_timesteps=250)
```

**Comment by EloyAnguiano** [comment]
My exact code NOW is this (this does not work): 

```python
import numpy as np
from stable_baselines3.common.env_util import make_vec_env
import gym
import grid2op
from grid2op.gym_compat import GymEnv, BoxGymObsSpace, DiscreteActSpace, BoxGymActSpace
from gym.spaces import Discrete, Box
from stable_baselines3 import PPO
from stable_baselines3.common.vec_env import DummyVecEnv, SubprocVecEnv


class GymEnvMultiProc(gym.Env):
    def __init__(self, gym_env_g2op):
        self.gym_env_g2op = gym_env_g2op
        self.action_space = Discrete(n=self.gym_env_g2op.action_space.n)
        self.observation_space = self.gym_env_g2op.observation_space
        
    def step(self, *args, **kwargs):
        return self.gym_env_g2op.step(*args, **kwargs)
    
    def reset(self, *args, **kwargs):
        return self.gym_env_g2op.reset(*args, **kwargs)
    
    def seed(self, *args, **kwargs):
        return self.gym_env_g2op.seed(*args, **kwargs)

    
def make_env(name):
    MAX_NODES = 40

    env = grid2op.make(name)
    gym_env = GymEnv(env)
    
    # Action Space
    gym_env.action_space = DiscreteActSpace(env.action_space, attr_to_keep=["set_bus"])

    # Observation space
    space_shape = (1,)
    observation = gym_env.observation_space
    observation.add_key(f"num_nodes",
                        lambda obs: obs.as_networkx().number_of_nodes(),
                        Box(shape=space_shape, 
                            low=np.full(space_shape, 0),
                            high=np.full(space_shape, MAX_NODES)))
    env_ok = GymEnvMultiProc(gym_env)
    return env_ok


if __name__ == '__main__':
    env = make_vec_env(make_env, env_kwargs = dict(
        name = "l2rpn_case14_sandbox"
    ),
    n_envs=1, vec_env_cls=SubprocVecEnv)
    model = PPO("MlpPolicy", env, verbose=1)
    model.learn(total_timesteps=250)
```

The key for tour approximation to run is that you redefine the observation space, but I need to expand the `Dict` observation space like the example with the `obs` object being your observation class for making the cast of the observation as a graph.

**Comment by BDonnot** [comment]
Hello,

I had a look at your code and unfortunately: 

1) I can't make it run with error: `AttributeError: Can't pickle local object 'make_env.<locals>.<lambda>'`  maybe it's not the code you used ? Have you tried this version ? I cannot help you if i can't manage to reproduce your issue unfortunately
2) I think you missed the point of the `GymEnvMultiProc` class. Indeed the observation space (and action space) should be a gym spaces only (and not something that inherit grid2op) so you should have something like in my example: 
```python
        self.observation_space = Box(low=self.gym_env_g2op.observation_space.low, 
                                     high=self.gym_env_g2op.observation_space.high)
```
I don't know exactly why you changed it but that is probably one of the cause of your trouble.

**Comment by BDonnot** [comment]
Hello,

After having spend some time to debug it (because yes, the bug was on the code and not on grid2op) here is a working example:

```python
import numpy as np
from stable_baselines3.common.env_util import make_vec_env
import gym
import grid2op
from grid2op.gym_compat import GymEnv, BoxGymObsSpace, DiscreteActSpace, BoxGymActSpace
from gym.spaces import Discrete, Box, Dict
from stable_baselines3 import PPO
from stable_baselines3.common.vec_env import DummyVecEnv, SubprocVecEnv
import copy
from stable_baselines3.common.env_checker import check_env


def complicated_fun(obs):
    return obs.as_networkx().number_of_nodes()
    
class GymEnvMultiProc(gym.Env):
    def __init__(self, gym_env_g2op):
        self.gym_env_g2op = gym_env_g2op
        self.action_space = Discrete(n=self.gym_env_g2op.action_space.n)
        self.observation_space = Dict(spaces=copy.deepcopy(self.gym_env_g2op.observation_space.spaces))
        
    def step(self, *args, **kwargs):
        return self.gym_env_g2op.step(*args, **kwargs)
    
    def reset(self, *args, **kwargs):
        return self.gym_env_g2op.reset(*args, **kwargs)
    
    def seed(self, *args, **kwargs):
        return self.gym_env_g2op.seed(*args, **kwargs)

def make_env(name):
    MAX_NODES = 40

    env = grid2op.make(name)
    gym_env = GymEnv(env)
    
    # Action Space
    gym_env.action_space = DiscreteActSpace(env.action_space, attr_to_keep=["set_bus"])

    # Observation space
    space_shape = (1,)
    observation = gym_env.observation_space
    observation.add_key(f"num_nodes",
                        lambda obs: np.array([obs.as_networkx().number_of_nodes()]),
                        Box(shape=space_shape, 
                            low=np.full(space_shape, fill_value=0),
                            high=np.full(space_shape, fill_value=MAX_NODES)))
    env_ok = GymEnvMultiProc(gym_env)
    return env_ok


if __name__ == '__main__':
    env_name = "l2rpn_case14_sandbox"
    
    # perform this check to make sure you can use it with sb3 !
    env = make_env(env_name)
    check_env(env)
    #####

    env = make_vec_env(make_env,
                       env_kwargs = dict(name=env_name),
                       n_envs=1,
                       vec_env_cls=SubprocVecEnv)
    model = PPO("MultiInputPolicy", env, verbose=1)
    model.learn(total_timesteps=250)
```

Another error came from a misusage of stable baselines 3. Next time, use their env_checker if you want to use their tools (see DLR-RM/stable-baselines3#1428 )

**Comment by EloyAnguiano** [comment]
It works now but only if I remove the `check_env` line. If I run the same code you provided I get:
```
Traceback (most recent call last):
  File "/home/eloy.anguiano/miniconda3/envs/l2rpn_new/lib/python3.10/site-packages/stable_baselines3/common/env_checker.py", line 219, in _check_returned_values
    _check_obs(obs[key], observation_space.spaces[key], "reset")
  File "/home/eloy.anguiano/miniconda3/envs/l2rpn_new/lib/python3.10/site-packages/stable_baselines3/common/env_checker.py", line 172, in _check_obs
    assert observation_space.contains(
AssertionError: The observation returned by the `reset()` method does not match the given observation space
```
Which libraries versions do you have?

**Comment by BDonnot** [comment]
Hello,

You can remove it indeed. This function just check for possible flaws in the environment.
In your case, looking at the error, it's probably a gym issue. Make sure to install a version 0.21 or something like that (see stable baselines 3 docs).

Let me know if something breaks again.


Ps. And I reiterate what I said before : you should really consider lightsim2grid as a backend for instant 15-20x speed up on one single core. Way more robust than all the multiprocess python shenanigans imho

**Comment by EloyAnguiano** [comment]
I removed It, but you said:

> Next time, use their env_checker if you want to use their tools 

So I wanted to put it on my pipeline for avoiding future errors. 

PD:  I am already isung the lightsim2grid on my main script and its going well, this one was a reproducible example so I did not specify that.

**Comment by BDonnot** [comment]
Hi,

Can I close this issue or is there anything you need to use this feature ?

**Comment by EloyAnguiano** [comment]
Still figuring out if this can cause the concurrency problem, but it seems to work. You can close it and then if I get that this was the problem when executing more than one agent at a time I will open a new issue. Thanks
